File: rbacpy/src/rbac.py
import logging
from dataclasses import dataclass
from typing import Generic, List

from common import Action, Authoritor, Permission, Role


@dataclass(frozen=True)
class RoleBasedAccess(Generic[Role], Authoritor[Role]):
    def is_authorized(self, permission: Permission[Role], role: Role, action: Action) -> bool:
        access = False
        perm: List[Role] = []

        match action:
            case Action.CREATE:
                logging.info("add create")
                perm.extend(permission.create_allowed_roles())
            case Action.READ:
                logging.info("add read")
                perm.extend(permission.read_allowed_roles())
            case Action.UPDATE:
                logging.info("add update")
                perm.extend(permission.update_allowed_roles())
            case Action.DELETE:
                logging.info("add delete")
                perm.extend(permission.delete_allowed_roles())
            case _:
                raise Exception("Unreachable")

        logging.info(f"{perm = }")

        if "*" in perm:
            access = True
        elif role in perm:
            access = True

        return access

File: rbac_rust/src/rbac.rs
use crate::common::{Authenticator, Permission, Action};

pub struct RoleBasedAccess;

impl Authenticator for RoleBasedAccess {
    fn is_authenticated<T: Permission>(&self, permission: &T, role: &T::Role, action: &Action) -> bool {
        let mut access = false;
        let mut perms: Vec<T::Role> = vec![];

        match action {
            Action::Create => {
                perms.extend(permission.create_allowed_roles());
            },
            Action::Read => {
                perms.extend(permission.read_allowed_roles());
            },
            Action::Update => {
                perms.extend(permission.update_allowed_roles());
            },
            Action::Delete => {
                perms.extend(permission.delete_allowed_roles());
            },
        }

        if perms.contains(&role) {
            access = true;
        }

        access
    }
}
